engine/src/main/java/org/camunda/bpm/engine/impl/util/xml/Parser.java

```diff
@@ -16,28 +16,80 @@
  */
 package org.camunda.bpm.engine.impl.util.xml;
 
+import org.camunda.bpm.engine.impl.ProcessEngineLogger;
+import org.camunda.bpm.engine.impl.cfg.ProcessEngineConfigurationImpl;
+import org.camunda.bpm.engine.impl.context.Context;
+import org.camunda.bpm.engine.impl.util.EngineUtilLogger;
+
+import javax.xml.XMLConstants;
 import javax.xml.parsers.SAXParser;
 import javax.xml.parsers.SAXParserFactory;
 
 
 /**
  * @author Tom Baeyens
  */
-public class Parser {
+public abstract class Parser {
 
-  protected static SAXParserFactory defaultSaxParserFactory = SAXParserFactory.newInstance();
-  
-  public static final Parser INSTANCE = new Parser();
+  protected static final EngineUtilLogger LOG = ProcessEngineLogger.UTIL_LOGGER;
 
-  public Parse createParse() {
-    return new Parse(this);
-  }
+  protected static final String EXTERNAL_GENERAL_ENTITIES = "http://xml.org/sax/features/external-general-entities";
+  protected static final String DISALLOW_DOCTYPE_DECL = "http://apache.org/xml/features/disallow-doctype-decl";
+  protected static final String LOAD_EXTERNAL_DTD = "http://apache.org/xml/features/nonvalidating/load-external-dtd";
+  protected static final String EXTERNAL_PARAMETER_ENTITIES = "http://xml.org/sax/features/external-parameter-entities";
+  protected static final String NAMESPACE_PREFIXES = "http://xml.org/sax/features/namespace-prefixes";
+
+  protected static ThreadLocal<SAXParserFactory> SAX_PARSER_FACTORY_INSTANCE = ThreadLocal.withInitial(SAXParserFactory::newInstance);
+
+  public abstract Parse createParse();
 
   protected SAXParser getSaxParser() throws Exception {
-    return getSaxParserFactory().newSAXParser();
+    SAXParserFactory saxParserFactory = getSaxParserFactoryLazily();
+    setXxeProcessing(saxParserFactory);
+    return saxParserFactory.newSAXParser();
+  }
+
+  protected SAXParserFactory getSaxParserFactoryLazily() {
+    return Parser.SAX_PARSER_FACTORY_INSTANCE.get();
+  }
+
+  protected void enableSchemaValidation(boolean enableSchemaValidation) {
+    SAXParserFactory saxParserFactory = getSaxParserFactoryLazily();
+    saxParserFactory.setNamespaceAware(enableSchemaValidation);
+    saxParserFactory.setValidating(enableSchemaValidation);
+
+    try {
+      saxParserFactory.setFeature(NAMESPACE_PREFIXES, true);
+    } catch (Exception e) {
+      LOG.unableToSetSchemaResource(e);
+    }
   }
 
-  protected SAXParserFactory getSaxParserFactory() {
-    return defaultSaxParserFactory;
+  protected void setXxeProcessing(SAXParserFactory saxParserFactory) {
+    Boolean enableXxeProcessing = isEnableXxeProcessing();
+    if (isEnableXxeProcessing() != null) {
```

Comment:
Is there any specific reason why you're not using the `enableXxeProcessing` variable in the if condition? You can avoid a second call to the `ProcessEngineConfiguration` this way. Also, the value of `enableXxeProcessing` would never be `null` since the default property value is `false`.

```suggestion
    if (enableXxeProcessing != null) {
```

Reply:
Nice catch! I overlooked it.